### ISSUE TITLE
Failure with name mangling when using "%" destruct pattern

### ISSUE THREAD

**Issue by RalfJung** [coq/coq#13413]
#### Description of the problem

To reproduce:
* Obtain commit ec3a2a2a0d70394da7e79e454ccd8c2fc344c935 of [Iris](https://gitlab.mpi-sws.org/iris/iris)
* Adjust `iris/prelude/prelude.v` to add `Global Set Name Mangling.`
* Go to this line:
https://gitlab.mpi-sws.org/iris/iris/-/blob/ec3a2a2a0d70394da7e79e454ccd8c2fc344c935/iris/proofmode/coq_tactics.v#L1051
```coq
destruct HΓp as [?|M C Γp ?%TCForall_Forall|? M C Γp Γp' []|? M Γp|M Γp]; simpl.
```
It fails saying 
```
Error: _1 is already used.
```
Removing the `%TCForall_Forall` makes the error disappear.

#### Coq Version

Commit 162c19cb

**Comment by SkySkimmer** [comment]
Can this be minimized any?

**Comment by RalfJung** [comment]
I mean, it probably can, but it'd be a lot of work to do this "black box" (based on how long it took last time, I expect 2h, maybe more). If you have an inkling for what might be the key factors contributing to the problem, that would help a lot.

**Comment by SkySkimmer** [comment]
I don't know much about destruct syntax. Maybe @herbelin @ppedrot ?

**Comment by herbelin** [comment]
The problem is that I don't know much about name mangling on my side. For instance, who is generating a name `_1`?

In any case, the code for `%TCForall_Forall` is `IntroApplyOn` in [tactics.ml](https://github.com/coq/coq/blob/0fc82e9651ee1dbc429c9b328b90ad8ad1a3cb14/tactics/tactics.ml#L2559). Should we look at `IntroNaming (IntroIdentifier id)` places? At `check_unicity`?

**Comment by SkySkimmer** [comment]
What's check_unicity?

**Comment by herbelin** [comment]
[check_name_unicity](https://github.com/coq/coq/blob/0fc82e9651ee1dbc429c9b328b90ad8ad1a3cb14/tactics/tactics.ml#L2430), sorry.

**Comment by jashug** [comment]
Here is a minimized example:
```coq
Set Mangle Names.
Goal forall (A B : Prop) (H : A -> B), A -> A -> B.
intros A B H ?%H _0. (* Fails with _0 is already used *)
```

**Comment by jashug** [comment]
@RalfJung If you want to try minimizing from the Iris example (which doesn't say `_1` anywhere, unlike my example; very suspicious), the problem is unlikely to depend on anything but the shape of the inductive you are eliminating (and probably just the problematic constructor), along with any `_n` names in context.

**Comment by RalfJung** [comment]
I guess let's see what it takes to fix your example, and if that helps with the problem in Iris.

**Comment by jashug** [comment]
@herbelin Anything that goes through `next_ident_away` or `next_ident_away_from` comes back as `_n`. That means both `NamingAvoid` and `NamingBasedOn` should get mangled, while `NamingMustBe` shouldn't.

**Comment by SkySkimmer** [comment]
>Here is a minimized example (or at least has similar symptoms):

Non mangle version:
~~~
Goal forall (A B : Prop) (H : A -> B), A -> A -> B.
intros A B H ?%H H0.
~~~

**Comment by herbelin** [comment]
Fixed in #13512. Thanks for the examples!
